File: VB/DashboardLocalizationManualCore/Models/CustomDashboardStorage.vb
Imports DevExpress.DashboardCommon
Imports DevExpress.DashboardWeb
Imports Microsoft.AspNetCore.Http
Imports System.Collections.Generic
Imports System.IO
Imports System.Linq
Imports System.Web
Imports System.Xml.Linq

Public Class CustomDashboardStorage
	Implements IEditableDashboardStorage

	Public Property RootPath() As String

	Public Sub New(ByVal path As String)
		RootPath = path
	End Sub

	Public Function GetAvailableDashboardsInfo() As IEnumerable(Of DashboardInfo)
		Dim dashboardInfos = New List(Of DashboardInfo)()

		Dim files = Directory.GetFiles(RootPath, "*.xml")

		For Each item In files
			Dim name = Path.GetFileNameWithoutExtension(item)

			If Not name.StartsWith("_") Then
				dashboardInfos.Add(New DashboardInfo() With {
					.ID = name,
					.Name = name
				})
			End If
		Next item

		Return dashboardInfos
	End Function

	Public Function LoadDashboard(ByVal dashboardID As String) As XDocument
		Dim path As System.String = System.IO.Path.Combine(RootPath, dashboardID & ".xml")
		Dim content = File.ReadAllText(path)

		Return XDocument.Parse(content)
	End Function

	Public Function AddDashboard(ByVal dashboard As XDocument, ByVal dashboardName As String) As String
'        var path = HttpContext.Current.Server.MapPath("~/App_Data/Dashboards/" + dashboardName + ".xml");
'
'        File.WriteAllText(path, dashboard.ToString());
'
'        return dashboardName;
		Return Nothing
	End Function

	Public Sub SaveDashboard(ByVal dashboardID As String, ByVal dashboard As XDocument)
'        var path = HttpContext.Current.Server.MapPath("~/App_Data/Dashboards/" + dashboardID + ".xml");
'
'        File.WriteAllText(path, dashboard.ToString());
	End Sub
End Class

File: CS/DashboardLocalizationCore/Models/CustomDashboardStorage.cs
using DevExpress.DashboardCommon;
using DevExpress.DashboardWeb;
using Microsoft.AspNetCore.Http;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Web;
using System.Xml.Linq;

public class CustomDashboardStorage : IEditableDashboardStorage {
    public string RootPath { get; set; }

    public CustomDashboardStorage(string path) {
        RootPath = path;
    }

    public IEnumerable<DashboardInfo> GetAvailableDashboardsInfo() {
        var dashboardInfos = new List<DashboardInfo>();

        var files = Directory.GetFiles(RootPath, "*.xml");

        foreach (var item in files) {
            var name = Path.GetFileNameWithoutExtension(item);

            if (!name.StartsWith("_"))
                dashboardInfos.Add(new DashboardInfo() { ID = name, Name = name });
        }

        return dashboardInfos;
    }

    public XDocument LoadDashboard(string dashboardID) {
        var path = Path.Combine(RootPath, dashboardID + ".xml");
        var content = File.ReadAllText(path);

        return XDocument.Parse(content);
    }

    public string AddDashboard(XDocument dashboard, string dashboardName) {
        /*var path = HttpContext.Current.Server.MapPath("~/App_Data/Dashboards/" + dashboardName + ".xml");

        File.WriteAllText(path, dashboard.ToString());

        return dashboardName;*/
        return null;
    }

    public void SaveDashboard(string dashboardID, XDocument dashboard) {
        /*var path = HttpContext.Current.Server.MapPath("~/App_Data/Dashboards/" + dashboardID + ".xml");

        File.WriteAllText(path, dashboard.ToString());*/
    }
}

